lib/spack/spack/environment/depfile.py

```diff
@@ -89,6 +90,32 @@ def accept(self, node):
         return True
 
 
+class MakefileSpec(object):
+    """Limited interface to spec to help generate targets etc. without
+    introducing unwanted special characters.
+    """
+
+    _pattern = None
+
+    def __init__(self, spec):
+        self.spec = spec
+
+    def safe_name(self):
+        return self.safe_format("{name}-{version}-{hash}")
+
+    def spec_hash(self):
+        return self.spec.dag_hash()
+
+    def safe_format(self, format_str):
+        unsafe_result = self.spec.format(format_str)
+        if not MakefileSpec._pattern:
+            MakefileSpec._pattern = re.compile(r"[^A-Za-z0-9_.-]")
+        return MakefileSpec._pattern.sub("_", unsafe_result)
```

Comment:
@haampie if I understand from our conversation this morning, you are saying this `re.sub` call could be moved into `Spec.format` itself (i.e. it would take another parameter like `filter_regex`, in which case that provides a way to resolve zlib's `=`-in-prefix issue). That is true but that consolidation doesn't seem to provide much benefit (i.e. moving that into `Spec` doesn't eliminate work on the caller's part other than that they wouldn't have to `import re` anymore).